spanner/tests/system/test_system.py

```diff
@@ -936,8 +936,86 @@ def test_read_w_ranges(self):
+        end_index = [data[end][0], data[end][1]]
+        closed_closed = KeyRange(start_closed=start_index,
+                                 end_closed=end_index)
+        closed_open = KeyRange(start_closed=start_index,
+                                 end_open=end_index)
+        open_open = KeyRange(start_open=start_index,
+                                 end_open=end_index)
+        open_closed = KeyRange(start_open=start_index,
+                                 end_closed=end_index)
+        range_type = ((closed_closed,),
+                      (closed_open,),
+                      (open_closed,),
+                      (open_open,)
+        )
+        indices = ((start, end+1),
+                   (start, end),
+                   (start+1, end+1),
+                   (start+1, end)
+        )
+        for index in range(len(range_type)):
+            keyset = KeySet(ranges=range_type[index])
+            rows = list(session.read(self.TABLE,
+                                     columns,
+                                     keyset,
+                                     index='name')
+            )
+            expected = data[indices[index][0]:indices[index][1]]
+            self.assertEqual(rows, expected)
+
+    def test_read_with_range_keys_index_with_limit(self):
+        row_count = 10
+        columns = self.COLUMNS[1], self.COLUMNS[2]
+        session, committed = self._set_up_table(row_count)
+        self.to_delete.append(session)
+        data = [[row[1], row[2]] for row in self._row_data(row_count)]
+        start = 3
+        end = 7
+        limit = 2
+        start_index = [data[start][0], data[start][1]]
+        end_index = [data[end][0], data[end][1]]
+        closed_closed = KeyRange(start_closed=start_index,
+                                 end_closed=end_index)
+        closed_open = KeyRange(start_closed=start_index,
+                                 end_open=end_index)
+        open_open = KeyRange(start_open=start_index,
+                                 end_open=end_index)
+        open_closed = KeyRange(start_open=start_index,
+                                 end_closed=end_index)
+        range_type = ((closed_closed,),
+                      (closed_open,),
+                      (open_closed,),
+                      (open_open,)
+        )
+        indices = ((start, end+1),
+                   (start, end),
+                   (start+1, end+1),
+                   (start+1, end)
+        )
+        for index in range(len(range_type)):
+            keyset = KeySet(ranges=range_type[index])
+            rows = list(session.read(self.TABLE,
+                                     columns,
+                                     keyset,
+                                     index='name',
+                                     limit=limit)
+            )
+            expected = data[indices[index][0]:indices[index][1]]
+            self.assertEqual(rows, expected[:limit])
```

Comment:
Rather than looping, please split these two test cases up into eight (four each, one per type of range).

Reply:
A simpler testcase is a) easier to review, and more importanly b) simpler to diagnose when it fails due to future changes.  Here is what one of them would look like:

```python
    def test_read_with_range_keys_index_closed_closed(self):
        row_count = 10
        columns = self.COLUMNS[1], self.COLUMNS[2]
        data = [[row[1], row[2]] for row in self._row_data(row_count)]
        session, _ = self._set_up_table(row_count)
        self.to_delete.append(session)
        start, end = 3, 7
        range = KeyRange(start_closed=data[start], end_closed=data[end])
        keyset = KeySet(ranges=(range,))
        rows = list(session.read(self.TABLE, columns, keyset, index='name'))
        self.assertEqual(rows, data[start:end + 1])
```